src/rc.js

```diff
@@ -25,57 +23,76 @@ const buildRcConf = () =>
 
     return values;
   });
-
-export function getRcConf(): {[string]: Array<string>} {
-  if (!rcConfCache) {
-    rcConfCache = buildRcConf();
-  }
-
-  return rcConfCache;
 }
 
-const buildRcArgs = () =>
-  Object.keys(getRcConf()).reduce((argLists, key) => {
-    const miniparse = key.match(/^--(?:([^.]+)\.)?(.*)$/);
+// get the built of arguments of a .yarnrc chain of the passed cwd
+function buildRcArgs(cwd: string): Map<string, Array<string>> {
+  const config = getRcConfigForCwd(cwd);
 
-    if (!miniparse) {
-      return argLists;
+  const argsForCommands: Map<string, Array<string>> = new Map();
+
+  for (const key in config) {
+    // args can be prefixed with the command name they're meant for, eg.
+    // `--install.check-files true`
+    const keyMatch = key.match(/^--(?:([^.]+)\.)?(.*)$/);
+    if (!keyMatch) {
+      continue;
     }
 
-    const namespace = miniparse[1] || '*';
-    const arg = miniparse[2];
-    const value = getRcConf()[key];
+    const commandName = keyMatch[1] || '*';
+    const arg = keyMatch[2];
+    const value = config[key];
 
-    if (!argLists[namespace]) {
-      argLists[namespace] = [];
-    }
+    // create args for this command name if we didn't previously have them
+    const args = argsForCommands.get(commandName) || [];
+    argsForCommands.set(commandName, args);
 
+    // turn config value into appropriate cli flag
     if (typeof value === 'string') {
-      argLists[namespace] = argLists[namespace].concat([`--${arg}`, value]);
+      args.push(`--${arg}`, value);
     } else if (value === true) {
-      argLists[namespace] = argLists[namespace].concat([`--${arg}`]);
+      args.push(`--${arg}`);
     } else if (value === false) {
-      argLists[namespace] = argLists[namespace].concat([`--no-${arg}`]);
+      args.push(`--no-${arg}`);
     }
+  }
 
-    return argLists;
-  }, {});
+  return argsForCommands;
+}
 
-export function getRcArgs(command: string): Array<string> {
-  if (!rcArgsCache) {
-    rcArgsCache = buildRcArgs();
+// extract the value of a --cwd arg if present
+function extractCwdArg(args: Array<string>): ?string {
+  for (let i = args.length - 1; i >= 0; i--) {
+    const arg = args[i];
+    if (arg === '--cwd') {
+      return args[i + 1];
+    }
   }
+  return null;
+}
```

Comment:
Not completely sure about this one. Running `yarn run -- --cwd foo` will break, the `--cwd` will be interpreted when it really shouldn't. Additionally, making a pseudo-parsing here means that it might get out of sync if we change the cli parsing in some unexpected ways (for example #4152).

I feel like a better option might be to do a first pass of cli parsing using the yarnrc options available in the current directory branch, then extract the `--cwd` option and, if present, run a second pass using the new directory as base. It would always be synced, and wouldn't have much perf impact in most cases. I don't know of feasable this is with commander tho.